core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/ast/HelperTool.java

```diff
@@ -21,13 +21,15 @@ public static void listToRdf(Collection<? extends Value> values, Resource head,
 		// The Turtle parser does not add "a rdf:List" statements when parsing the shorthand list format,
 		// so we don't add rdf:List when writing it out either.
 
+		int counter = 0;
+
 		Iterator<? extends Value> iter = values.iterator();
 		while (iter.hasNext()) {
 			Value value = iter.next();
 			model.add(head, RDF.FIRST, value);
 
 			if (iter.hasNext()) {
-				Resource next = bnode();
+				Resource next = bnode(head.stringValue() + "_rdf_list_item_" + (++counter));
```

Comment:
Here we mint stable identifiers for the blank nodes based on the identifier of the first blank node in the list.

Reply:
Can you use a better class name than "HelperTool"? Also: this method duplicates existing functionality for creating RDF collections from a Java lists. Any reason you're rolling your own instead of just reusing/improving/extending the existing utilities?

Reply:
I think I was initially struggling with pretty printing. Then I also found out I wanted type coercion. And now I've found out I want stable identifiers for each blank node in the list.

Reply:
Does our one use stable blank node identifiers for each item in the list?

Reply:
I'm not sure. But this all sounds potentially generally useful, so why not improve the existing utilities instead of adding a duplicate?

Reply:
Also depends on what exactly you mean by stable, of course.

Reply:
The reason for this is that the toModel method may be called multiple times with the same model. So it's important that all identifiers are stable so we don't create a new list every time. 

I think it's a bad idea to add this to the general rdf list code.

Reply:
> The reason for this is that the toModel method may be called multiple times with the same model. So it's important that all identifiers are stable so we don't create a new list every time.

So... correct me if I'm wrong, you are using stable node identifiers so that if the same list of items is added to the output model twice, it does not result in a second list if there already is one (because they will merge based on bnode identity)? I don't understand why you're solving that in this fashion instead of just quickly checking if there is an existing list or not before adding.

> I think it's a bad idea to add this to the general rdf list code.

This specific feature probably, but apart from being uncomfortable with the use to which you put this here, I'm left wondering if there is a better way to do it (e.g. some way we can let a user inject a lambda into the generic utility which can handle these kinds of bespoke conversions). I also think that the type coercion you mentioned might be relevant more generally.

Refactoring all of this is probably out of scope for this fix, but at the very least, I think we should make sure that:

1. `HelperTool` is renamed (to something meaningful like `ShaclRdfListUtil` perhaps);
2. It should be package-private or if that is not technically possible, marked as `@InternalUseOnly`;
3. It needs to clearly have the assumptions you're making in the javadoc (so that it is clear why we need this bespoke rdf collection function instead of the general one).